plugins/ata/fu-ata-device.c

```diff
@@ -0,0 +1,603 @@
+	guint32 block_count = data_sz / FU_ATA_BLOCK_SIZE;
+	guint32 buffer_offset = addr / FU_ATA_BLOCK_SIZE;
+
+	/* write block */
+	tf.dev = 0xa0 | ATA_USING_LBA;
+	tf.command = ATA_OP_DOWNLOAD_MICROCODE;
+	tf.feat = self->transfer_mode;
+	tf.nsect = block_count & 0xff;
+	tf.lbal = block_count >> 8;
+	tf.lbam = buffer_offset & 0xff;
+	tf.lbah = buffer_offset >> 8;
+	if (!fu_ata_device_command (self, &tf, SG_DXFER_TO_DEV,
+				    120 * 1000, /* a long time! */
+				    (guint8 *) data, data_sz, error)) {
+		g_prefix_error (error, "failed to write firmware @0x%0x",
+				(guint) addr);
+		return FALSE;
+	}
+
+	/* check drive status */
+	if (tf.nsect == 0x0)
+		return TRUE;
+
+	/* drive wants more data, or thinks it is all done */
+	if (tf.nsect == 0x1 || tf.nsect == 0x2)
+		return TRUE;
+
+	/* the offset was set up incorrectly */
+	if (tf.nsect == 0x4) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "alignment error");
+		return FALSE;
+	}
+
+	/* other error */
+	g_set_error (error,
+		     G_IO_ERROR,
+		     G_IO_ERROR_INVALID_DATA,
+		     "unknown return code 0x%02x",
+		     tf.nsect);
+	return FALSE;
+}
+
+static gboolean
+fu_ata_device_write_firmware (FuDevice *device, GBytes *fw, GError **error)
+{
+	FuAtaDevice *self = FU_ATA_DEVICE (device);
+	g_autoptr(GPtrArray) chunks = NULL;
+	guint max_size = 0xffff * FU_ATA_BLOCK_SIZE;
+
+	/* only one block allowed */
+	if (self->transfer_mode == 0x7)
+		max_size = 0xffff;
+
+	/* check is valid */
+	if (g_bytes_get_size (fw) > max_size) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "firmware is too large, maximum size is %u",
+			     max_size);
+		return FALSE;
+	}
+	if (g_bytes_get_size (fw) % FU_ATA_BLOCK_SIZE != 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "firmware is not multiple of block size %i",
+			     FU_ATA_BLOCK_SIZE);
+		return FALSE;
+	}
+
+	/* build packets */
+	chunks = fu_chunk_array_new_from_bytes (fw,
+						0x00,		/* start_addr */
+						0x00,		/* page_sz */
+						self->transfer_size);
+
+	/* write each block */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_ata_device_fw_download (self,
+						chk->idx,
+						chk->address,
+						chk->data,
+						chk->data_sz,
+						error)) {
+			g_prefix_error (error, "failed to write chunk %u: ", i);
+			return FALSE;
+		}
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len + 1);
+	}
+
+	/* success! */
+	fu_device_set_progress (device, 100);
+	return TRUE;
+}
+
+static gboolean
+fu_ata_device_set_quirk_kv (FuDevice *device,
+			    const gchar *key,
+			    const gchar *value,
+			    GError **error)
+{
+	FuAtaDevice *self = FU_ATA_DEVICE (device);
+	if (g_strcmp0 (key, "AtaTransferMode") == 0) {
+		self->transfer_mode = fu_common_strtoull (value);
```

Comment:
Since you only can support `0x3`, `0x7` or `0xe` shouldn't these be guarded here?

Reply:
Fixed, agreed.